File: Go/card.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

type card struct {
	done  bool
	mark  string
	num   byte
	value byte
}

func (c *card) format() string {
	return fmt.Sprintf("[%s%d]", c.mark, c.num)
}

type cardDeck struct {
	cards [52]card
}

func (cd *cardDeck) init() {
	rand.Seed(time.Now().UnixNano())
	i := 0

	for _, mark := range [4]string{"S", "H", "D", "C"} {
		for n := 1; n <= 13; n++ {
			cd.cards[i] = card{
				done:  false,
				mark:  mark,
				num:   byte(n),
				value: byte(math.Min(float64(n), 10)),
			}
			i++
		}
	}
}

func (cd *cardDeck) pick() (card *card) {
	for card == nil || card.done {
		rnd := rand.Intn(52)
		card = &cd.cards[rnd]
	}

	card.done = true
	return card
}

File: Rust/src/card.rs
use rand::{thread_rng, Rng};

#[derive(Debug)]
pub struct Card {
    mark: char,
    num: u8,
}

impl Card {
    pub fn disp(&self) -> String {
        format!("{}{}", self.mark, self.num)
    }

    pub fn val(&self) -> u8 {
        self.num.min(10)
    }
}

pub struct Deck {
    cards: Vec<Card>,
}

impl Deck {
    /** デッキ作成 */
    pub fn init() -> Self {
        let marks = ['C', 'D', 'H', 'S'];
        let mut cards: Vec<Card> = Vec::new();

        for i in 0..52 {
            let num = (i % 13 + 1) as u8;
            let card = Card {
                mark: marks[i % 4],
                num,
            };
            cards.push(card);
        }

        Self { cards }
    }

    /** デッキから1枚ひく */
    pub fn pick(&mut self) -> Card {
        let index = thread_rng().gen_range(0..self.cards.len());
        self.cards.remove(index)
    }
}
